Allow combining EL expression with interpolated values

Currently a statement like 

```
- IF [${$[env:MY_VARIABLE;default='default'] ne ''}]
```

leads to the exception

```
Caused by: javax.el.ELException: Failed to parse the expression [${$[env:MY_VARIABLE;default='default'] ne ''}]
	at org.apache.el.lang.ExpressionBuilder.createNodeInternal(ExpressionBuilder.java:148)
	at org.apache.el.lang.ExpressionBuilder.build(ExpressionBuilder.java:178)
	at org.apache.el.lang.ExpressionBuilder.createValueExpression(ExpressionBuilder.java:242)
	at org.apache.el.ExpressionFactoryImpl.createValueExpression(ExpressionFactoryImpl.java:67)
	at biz.netcentric.cq.tools.actool.configreader.YamlMacroElEvaluator.evaluateEl(YamlMacroElEvaluator.java:102) [biz.netcentric.cq.tools.accesscontroltool.bundle:2.7.0]
	at biz.netcentric.cq.tools.actool.configreader.YamlMacroProcessorImpl.evaluateIfStatement(YamlMacroProcessorImpl.java:241) [biz.netcentric.cq.tools.accesscontroltool.bundle:2.7.0]
	at biz.netcentric.cq.tools.actool.configreader.YamlMacroProcessorImpl.transform(YamlMacroProcessorImpl.java:142) [biz.netcentric.cq.tools.accesscontroltool.bundle:2.7.0]
	at biz.netcentric.cq.tools.actool.configreader.YamlMacroProcessorImpl.transform(YamlMacroProcessorImpl.java:122) [biz.netcentric.cq.tools.accesscontroltool.bundle:2.7.0]
	at biz.netcentric.cq.tools.actool.configreader.YamlMacroProcessorImpl.processMacros(YamlMacroProcessorImpl.java:64) [biz.netcentric.cq.tools.accesscontroltool.bundle:2.7.0]
	at biz.netcentric.cq.tools.actool.configreader.YamlConfigurationMerger.getMergedConfigurations(YamlConfigurationMerger.java:134) [biz.netcentric.cq.tools.accesscontroltool.bundle:2.7.0]
	at biz.netcentric.cq.tools.actool.impl.AcInstallationServiceImpl.installConfigurationFiles(AcInstallationServiceImpl.java:292) [biz.netcentric.cq.tools.accesscontroltool.bundle:2.7.0]
	at biz.netcentric.cq.tools.actool.impl.AcInstallationServiceImpl.installConfigurationFiles(AcInstallationServiceImpl.java:261) [biz.netcentric.cq.tools.accesscontroltool.bundle:2.7.0]
```

So it seems that interpolation cannot be used inside EL now. We should lift this limitation.

--- a/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigurationAdminPluginScalarConstructor.java
+++ b/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigurationAdminPluginScalarConstructor.java
@@ -54,6 +54,7 @@ public class YamlConfigurationAdminPluginScalarConstructor extends Constructor {
         dictionary.put(KEY, scalar);
         try {
             installLog.addMessage(LOG, "Using ConfigAdminPlugin to resolve '" + scalar + "'");
+            // resolve arbitrarily many placeholders in the scalar
             interpolationPlugin.modifyConfiguration(null, dictionary);
             return (String)dictionary.get(KEY);
         } catch (Exception e) {

--- a/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigurationMerger.java
+++ b/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigurationMerger.java
@@ -82,7 +82,7 @@ public class YamlConfigurationMerger implements ConfigurationMerger {
     ConfigurationPlugin interpolationPlugin;
 
     /** Regular expression which matches against values which should be interpolated, see https://github.com/apache/felix-dev/blob/master/configadmin-plugins/interpolation/README.md */
-    public static final Pattern CONFIG_ADMIN_INTERPOLATOR_FORMAT = Pattern.compile("^\\$\\[(env|secret|prop):.*\\]$");
+    public static final Pattern CONFIG_ADMIN_INTERPOLATOR_FORMAT = Pattern.compile(".*\\$\\[(env|secret|prop):[^\\]]*\\].*");
 
     @Override
     public AcConfiguration getMergedConfigurations(
@@ -104,7 +104,7 @@ public class YamlConfigurationMerger implements ConfigurationMerger {
         if (interpolationPlugin != null) {
             yamlParser = new Yaml(new YamlConfigurationAdminPluginScalarConstructor(installLog, interpolationPlugin));
             // bind constructor to certain scalar formats (compare with https://bitbucket.org/asomov/snakeyaml/src/master/src/test/java/org/yaml/snakeyaml/env/EnvVariableTest.java)
-            yamlParser.addImplicitResolver(YamlConfigurationAdminPluginScalarConstructor.TAG, CONFIG_ADMIN_INTERPOLATOR_FORMAT, "$[" );
+            yamlParser.addImplicitResolver(YamlConfigurationAdminPluginScalarConstructor.TAG, CONFIG_ADMIN_INTERPOLATOR_FORMAT, null);
             installLog.addMessage(LOG, "Using YAML parser with ConfigurationAdmin Plugin placeholder support");
         } else {
             yamlParser = new Yaml();

--- a/accesscontroltool-bundle/src/test/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigurationMergerTest.java
+++ b/accesscontroltool-bundle/src/test/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigurationMergerTest.java
@@ -120,11 +120,15 @@ public class YamlConfigurationMergerTest {
             // replace all placeholders with value "resolved"
             @Override
             public void modifyConfiguration(ServiceReference<?> reference, Dictionary<String, Object> properties) {
-                properties.put(YamlConfigurationAdminPluginScalarConstructor.KEY, "resolved");
+                // replace all placeholders
+                String value = (String)properties.get(YamlConfigurationAdminPluginScalarConstructor.KEY);
+                String newValue = value.replaceAll("\\$\\[(env|secret|prop):([^\\]]*)\\]", "resolved-$2");
+                properties.put(YamlConfigurationAdminPluginScalarConstructor.KEY, newValue);
             }
         };
         AcConfiguration acConfiguration = getAcConfigurationForFile(yamlConfigMerger, session, "test-placeholders.yaml");
-        assertEquals("resolved", acConfiguration.getAuthorizablesConfig().getAuthorizableConfig("editor").getPassword());
+        assertEquals("resolved-password", acConfiguration.getAuthorizablesConfig().getAuthorizableConfig("editor").getPassword());
+        assertEquals("RESOLVED-PASSWORD", acConfiguration.getAuthorizablesConfig().getAuthorizableConfig("editor2").getPassword());
     }
 
     public static AcConfiguration getAcConfigurationForFile(YamlConfigurationMerger merger, Session session, String testConfigFile)

--- a/accesscontroltool-bundle/src/test/resources/test-placeholders.yaml
+++ b/accesscontroltool-bundle/src/test/resources/test-placeholders.yaml
@@ -9,3 +9,5 @@
 - user_config:
   - editor:
     - password: $[secret:password]
+  - editor2:
+    - password: ${upperCase('$[secret:password]')}

